fix(forms): improve select performance and ensure that selected option matches the ngModel value

theodorejb opened the pull request:
Fixes #41330, fixes #14505, fixes #18430.

## PR Checklist
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?
- [x] Bugfix

## What is the current behavior?
Currently Angular calls the `writeValue()` over and over for every individual option element that is added or removed from the DOM. This results in exponentially more `_compareValue` calls than the number of option elements (issue #41330).

Furthermore, calling `writeValue()` when rendering individual option elements causes an issue in Safari and IE 11 where the first option element fails to be deselected when no option matches the select ngModel. This is because Angular sets the select element's value property before appending the option's child text node to the DOM (issue #14505).

Finally, the current behavior causes an issue with delayed element removal when using the animations module (in all browsers). When a selected option is removed (so no option matches the ngModel anymore), Angular changes the select element value before actually removing the option from the DOM. Then when the option is finally removed from the DOM, the browser changes the select value to that of the first option, even though it doesn't match the ngModel (issue #18430).

## What is the new behavior?
Rather than calling `writeValue()` over and over for individual option elements that are added/removed from the DOM, call it once after all the options have finished rendering via the `afterNextRender` function. This dramatically improves performance and results in correct behavior across all supported browsers.

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

theodorejb commented:
@kara Would it be possible for you to review this soon?

trotyl commented:
Ref https://github.com/angular/angular/pull/14507

theodorejb commented:
@trotyl I don't think pull request #14507 was ever finished, since it didn't fix the issue for me when I tested it.

trotyl commented:
That's why “ref”, not “duplicate of”.

JoostK commented:
I experience the same broken behaviour in Safari 11.0.2

theodorejb commented:
I just tested https://angular-select-ie-bug.stackblitz.io/ in Safari on the latest version of iOS and can confirm that the bug happens there as well.

theodorejb commented:
@kara This is still an issue in Angular 7.0.0-rc.1, affecting the latest versions of both IE and Safari. Could you please let me know if there's anything that I need to change, or additional tests I should add before this PR can be merged?

kriks57 commented:
@kara please have a look, and we would also need this fix in angular 4

tan9 commented:
@kara , @jasonaden, @IgorMinar is there anything I can do or @theodorejb can do to make progress on the triage or reviewing this merge request? I am willing to help.

dasaweb commented:
> @kara , @jasonaden, @IgorMinar is there anything I can do or @theodorejb can do to make progress on the triage or reviewing this merge request? I am willing to help.

This problem has been around for years now, and it's still a big problem for me (and for others, too, see https://github.com/angular/angular/issues/14505). It would be very helpful if the proposed solution could be incorporated. Thanks a lot!

theodorejb commented:
@AndrewKushnir Can you please review this? It would be really nice if this bug was fixed in Angular 9.

mhevery commented:
[presubmit](http://test/OCL:275273819:BASE:275273874:1571329467954:4241db52)

AndrewKushnir commented:
Thanks for the PR @theodorejb. Could you please rebase it on top of the latest master, so CI runs again with the most up-to-date set of tests?

It's be also helpful if you could provide some additional information regarding the problem:
- if you tried to explore other solutions to this issue (that potentially would not require `setTimeout`)
- if this issue is reproducible with plain HTML and JS, so we can make sure it's a specific browser behavior that is triggering this issue
- if the problem is reproducible with HTML and JS (i.e. it's not Angular-specific), there might be existing bug reports created for these browsers. It'd be helpful to have these links, so we can get extra information and extend comments in the code

Thank you.

theodorejb commented:
Thank you so much for the review @AndrewKushnir. The PR has been rebased on master now and a couple code comments have been improved.

I spent a bunch of time digging through the `select_control_value_accessor.ts` code and setting debugger breakpoints to figure out why it behaves the way it does.

When option elements are rendered, Angular calls `writeValue` on the select component after setting the value of each option element. The problem occurs because `writeValue` is called for each option **before that option's child text node has been added to the DOM.** Hence `value` and `selectedIndex` are set on the select element before the option has a text node, and when the text node is finally added,  Safari and IE ignore the preveously set `selectedIndex`.

I made a simple JS Bin demonstrating the issue with plain HTML and JavaScript:
https://jsbin.com/vidatokizo/1/edit?html,js,output. In Chrome, Firefox, and Edge no option is selected. In Safari and IE 11 the first option is selected.

The reason that using `setTimeout` fixes the issue is because it causes `selectedIndex` to be set *after* the option child text node has been added, in which case it correctly applies in all browsers.

The only other way to fix the issue would be to somehow ensure that `writeValue` isn't called on the select component until after the option's text node has been appended. However, I'm not familiar enough with Angular's internals to know how to accomplish this (and it seems like it could require bigger hacks than `setTimeout`).

My reasoning is that at least this approach fixes the issue and adds tests to ensure it isn't regressed. Later on it will always be possible to improve the code with a larger/more complex refactor.

Please let me know if there's any further information I can provide to help move this forward.

theodorejb commented:
@AndrewKushnir can you review this again? I simplified the code further, moving the call to set the select element's value property inside the timeout and completely removing the need to conditionally set `selectedIndex`. I also updated one of the tests to add a `tick()` call like most of the other tests have. That particular test only incidentally worked in Internet Explorer previously without the `tick()` call since its option elements are already rendered in the DOM with their child text nodes.

AndrewKushnir commented:
Hi @theodorejb, thanks for additional investigation and information.

We are in the process of getting v9 ready and we'll get back to this issue after v9 is released. I've added the "PR action: review" to this PR for now.

Thank you.

theodorejb commented:
Recently I ran into issue #18430 (affecting all browsers) where select dropdowns go out of sync with their ngModel when `BrowserAnimationsModule` is imported (due to element removal being delayed until after change detection). This pull request also fixes that bug.

Additionally, this pull request can significantly improve performance when rendering dropdowns with hundreds or thousands of options (since `_getOptionId` and `_buildValueString` now only run once per change, rather than over and over for every single option element that is added).

theodorejb commented:
@AndrewKushnir Can you please review this again now that v9 has been released?

theodorejb commented:
@mhevery Is there anything else I can do to help get this reviewed/merged? The issue keeps leading to bugs in our app which require ugly hacks to work around.

Also this PR can noticeably improve performance for select dropdowns with a lot of options. Can someone on the Angular team please review it?

I hate to be a bother, but I've been trying to get this PR reviewed for almost two years now.

@kara @AndrewKushnir @petebacondarwin @IgorMinar

petebacondarwin commented:
@theodorejb - I have pinged Andrew directly about this since he was the last person in the Angular team commenting on this.

AndrewKushnir commented:
HI @theodorejb, sorry for the delay in answering.

I've added this PR into the `v10-candidates` Milestone and it will be prioritized as a part of working on v10. I don't have any ETAs at this moment, but we'll keep this thread updated once we have more info.

Thank you.

xonuzofa commented:
Please merge this!  
Also here: [Issue 35896](https://github.com/angular/angular/issues/35896)

theodorejb committed:
fix(forms): improve select performance and ensure that selected option matches the ngModel value

Rather than calling `writeValue()` over and over for individual option elements that are
added/removed from the DOM, call it once after all the options have finished rendering
via the `ngAfterViewChecked` hook.

This is needed to efficiently set the select value when adding/removing options. The
previous approach resulted in exponentially more `_compareValue` calls than the number
of option elements (issue #41330).

Furthermore, calling `writeValue()` when rendering individual option elements caused
an issue in Safari and IE 11 where the first option element failed to be deselected when
no option matched the select ngModel. This was because Angular set the select element's
value property before appending the option's child text node to the DOM (issue #14505).

Finally, this PR fixes an issue with delayed element removal when using the animations
module (in all browsers). Previously when a selected option was removed (so no option
matched the ngModel anymore), Angular changed the select element value before actually
removing the option from the DOM. Then when the option was finally removed from the DOM,
the browser would change the select value to that of the first option, even though it
didn't match the ngModel (issue #18430).

Fixes #41330, fixes #14505, fixes #18430.

theodorejb commented:
@AndrewKushnir I noticed that the `v10-candidates` milestone was closed a few days ago, and 10.0.0-rc.0 has been published. Can this PR please still be prioritized for v10? Many forms with select dropdowns are broken right now because of this bug.

The PR is pretty simple, self-contained, and (I think) well-documented, so it shouldn't take long to review. Thanks!

AndrewKushnir commented:
Hi @theodorejb,

Sorry for the delay in answering.

We've discussed this change with Kara and the conclusion is that making the logic async will be a breaking change affecting a lot of cases. As a possible solution, it might be great to explore either a fully synchronous solution or a hybrid approach where:
- the value is set synchronously (as we do it now)
- check what the actual value is after a timeout (via `setTimeout` or may be better using `Promise.resolve` to have it in a microtask)
- if the value is different (which indicates that there is a problem) - apply the necessary one once again
- if the value is the same - do nothing

What do you think?

Thank you.

theodorejb commented:
@AndrewKushnir Thanks so much for the review. I updated the PR now to use the hybrid approach you suggested to avoid any potential backwards compatibility break. I hope this can make it into Angular 10.

Longer term it would be ideal to have a fully synchronous solution, where the select value wouldn't be set until after all option elements are rendered with their text nodes. Such an approach has the potential to noticeably improve the performance of select dropdowns with a lot of options. However, it's a bit beyond my current knowledge of Angular internals to implement - it seems like it would require deeper hooks into the Angular renderer and animations module.

AndrewKushnir commented:
Hi @theodorejb, thanks for updating the PR!

We performed some additional investigation today (thanks @atscott for the help) to look for a synchronous solution to avoid the risks of unexpected value reset after a timeout (since we'll do that in a subsequent change detection cycle). It looks like `ngAfterViewChecked` lifecycle hook can be defined in the `SelectControlValueAccessor` class to trigger the logic to reconcile the DOM with the data (set `selectedIndex` for the `<select>` element). Here is a draft version:

```
  ngAfterViewChecked() {
    const id = this._getOptionId(this.value);
    const nativeElement = this._elementRef.nativeElement;
    // TODO: this code needs a detailed comment explaining why we do that
    if (id == null && nativeElement.selectedIndex !== -1) {
      this._renderer.setProperty(nativeElement, 'selectedIndex', -1);
    }
  }
```
I did a quick test with the cases that you added in this PR and they seem to work correctly. Note: no changes are required in the `writeValue` function in this case and the change is pretty isolated, so it'd be easy to clean it up once the problem is resolved in Safari and IE 11 is deprecated (at some point in the future).

Please let me know what do you think about the proposed solution.

Thank you.

theodorejb commented:
Hi @AndrewKushnir,

I added a test now to ensure the animations module issue isn't regressed.

Initially I tried putting this test in the same file as the other select tests, but when I added `BrowserAnimationsModule` as an import there, the test failed with "Error: The selector "#root201" did not match any elements". Apparently the forms test components are rendered with `TestComponentRenderer`, which doesn't insert the root element (so it breaks when the animations module tries to select the root element). Perhaps the forms tests could be changed to use `DOMTestComponentRenderer` instead, but I wasn't able to figure out how to do this. Hopefully it's fine to just keep the test with other animation module tests.

After adding the test, I tried the `ngZone.onStable` approach you suggested. Indeed, it does seem to resolve the issue with the animations module. However, it also broke a couple other forms tests:

```
HeadlessChrome 83.0.4103 (Windows 10.0.0) value accessors select controls in reactive forms should support primitive values FAILED
        Error: Expected '' to equal 'SF'.
            at <Jasmine>
            at UserContext.<anonymous> (packages/forms/test/value_accessor_integration_spec.ts:192:45 <- bin/packages/forms/test/value_accessor_integration_spec.js:168:60)


HeadlessChrome 83.0.4103 (Windows 10.0.0) value accessors select controls in template-driven forms when there is a placeholder option FAILED       
        Error: Expected false to be true.
            at <Jasmine>
            at UserContext.<anonymous> (packages/forms/test/value_accessor_integration_spec.ts:333:56 <- bin/packages/forms/test/value_accessor_integration_spec.js:291:68)
```

I wasn't able to figure out how to fix these, so for now this PR still implements the hybrid approach with `setTimeout`, which doesn't break anything.

If it's not easy to find another approach that passes all the tests, I hope the current patch can still be merged. It will always be possible to improve the implementation further in the future if a better approach is found (and tests will be in place to make sure everything keeps working).

Thanks!

AndrewKushnir commented:
Hi @theodorejb, thanks for adding extra tests and providing more details!

I think it'd be great to investigate why these two tests started to fail, so we have the full picture.

The reason why I'm not very comfortable with adding the `setTimeout` is that it would trigger a new Change Detection cycle, which is not needed for non-Safari and non-IE users (~19.5% of all users) and having a check for these specific browsers in the code (like `if (IE || Safari) ...`) would introduce a potentially inconsistent behavior across browsers (1 CD cycle in one set of browsers and 2 in another). The `ngZone.onStable` solution seems like the right balance (extra logic to verify the state of the DOM, but without triggering potentially expensive CD cycle), so I'd be great to see why it doesn't work for some of the cases.

Please let us know if you could perform further investigation and/or need help with it.

Thank you.

theodorejb commented:
Hi @AndrewKushnir, I figured out why the two tests failed. The problem was that `selectedIndex` was being set rather than the select value. This broke the placeholder case, where the Angular option ID is `null` but the manually entered placeholder option should still be selected. It also broke the reactive forms test because it doesn't use `ngModel` (so the option ID is null again).

Instead of setting `selectedIndex` I changed the hook to simply call `writeValue`, and now all the tests are passing. :)

Let me know if any other changes are needed. I really appreciate your help in moving this forward!

theodorejb commented:
The test_aio and test_aio_local test failures seem unrelated, as they are also happening on master.

AndrewKushnir commented on packages/forms/src/directives/select_control_value_accessor.ts:
```diff
@@ -123,20 +123,37 @@ export class SelectControlValueAccessor implements ControlValueAccessor {
+     * the select value to that of the first option, even though it doesn't match the ngModel.
+     */
+
+    this._ngZone.onStable.pipe(take(1)).subscribe(() => {
```
It looks like `_ngZone` is set as `null` in the tests in the `packages/forms/test/directives_spec.ts` file, wouldn't it throw an error here?

theodorejb replied:
The other two params (`_renderer` and `_elementRef`) are also set to `null` there, but they don't throw.

AndrewKushnir commented on packages/forms/src/directives/select_control_value_accessor.ts:
```diff
@@ -123,20 +123,37 @@ export class SelectControlValueAccessor implements ControlValueAccessor {
+  ngAfterViewChecked() {
+    /*
+     * When no option value matches the select value, in order for the first option to be
+     * deselected in Safari and Internet Explorer, the select element's value property must
```
Could you please also mention which versions of Safari and IE are affected? That should help cleanup this logic once corresponding browser versions are deprecated in case of IE. For Safari, that should help keep track of the major browser version where the problem was present and if it's resolved at some point, we'll be able to identify the deprecation period before removing this logic (in case it's no longer needed).

theodorejb replied:
I documented the versions I've tested with. But probably the logic will have to stay regardless of whether Safari's behavior ever changes since it's also needed to fix the issue when using the animations module.

AndrewKushnir commented on packages/forms/src/directives/select_control_value_accessor.ts:
```diff
@@ -123,20 +123,37 @@ export class SelectControlValueAccessor implements ControlValueAccessor {
+     */
+
+    this._ngZone.onStable.pipe(take(1)).subscribe(() => {
+      this.writeValue(this.value);
```
I think it'd be great to do reconciliation of DOM and the current state only when the discrepancy is detected (similar to what I had in the example code in one of my previous comments).

theodorejb replied:
I could do that, but it will require storing the last set string value in a new property. I don't think it hurts to call `writeValue` again even when it won't change anything - this already happens for every single option element that is rendered.

theodorejb replied:
This change is done now. Personally I'm not convinced it's better than just calling `writeValue` even when it's not necessary, but we can easily change this again if needed depending on what you think.

AndrewKushnir commented on packages/forms/src/directives/select_control_value_accessor.ts:
```diff
@@ -123,20 +123,37 @@ export class SelectControlValueAccessor implements ControlValueAccessor {
   writeValue(value: any): void {
     this.value = value;
     const id: string|null = this._getOptionId(value);
-    if (id == null) {
```
We should probably keep this logic to make sure we are not changing the behavior. Please let me know if I'm missing something.

theodorejb replied:
This does not change behavior in any way, since the `selectedIndex` is immediately overridden by setting the select value.

AndrewKushnir replied:
Thanks for additional clarification @theodorejb. If this doesn't affect the logic in this PR, I think we should keep it and create a separate ticket (and a separate PR) to remove it, so that we keep only relevant changes in this PR and avoid any risks related to removing that code.

theodorejb replied:
Okay, I created a separate PR for this here: https://github.com/angular/angular/pull/37614

AndrewKushnir commented on packages/forms/src/directives/select_control_value_accessor.ts:
```diff
@@ -228,6 +245,8 @@ export class NgSelectOption implements OnDestroy {
     if (this._select == null) return;
     this._select._optionMap.set(this.id, value);
     this._setElementValue(_buildValueString(this.id, value));
+    // TODO: running this for every option element is bad for performance.
```
Thanks for this noticing that, I do agree that we should look at the possible solutions to improve the performance. Could you please create a new ticket and describe it there (+ may be describe some possible solutions or other details if you had a chance to investigate this)? I think having a ticket should be enough and we can avoid adding this comment here.

theodorejb replied:
I removed these comments now and can file a new issue when I have time.

AndrewKushnir commented on packages/platform-browser/animations/test/animation_renderer_spec.ts:
```diff
@@ -322,9 +324,50 @@ describe('AnimationRendererFactory', () => {
     expect(renderer.log).toEqual(['begin', 'end']);
   });
+
+  it('should not break select controls when options are removed', fakeAsync(() => {
```
I think this test should be in the `forms` package. You mentioned that there are some errors that appeared when you tried adding it there. We can look into that once the main logic is finalized.

theodorejb replied:
I removed this test for now, since the added `expect(select.nativeElement.value).not.toBe('0: Object');` in value_accessor_integration_spec.ts seems to cover the same thing (it also fails if `NgZone.onStable` is removed, and this is the same expectation that would fail in the removed test).

AndrewKushnir commented on packages/forms/src/directives/select_control_value_accessor.ts:
```diff
@@ -123,20 +123,37 @@ export class SelectControlValueAccessor implements ControlValueAccessor {
+     * rendering options Angular calls the select component's writeValue() method before
+     * adding text nodes to the DOM, so it is necessary to call it again after checking the view.
+     *
+     * This is also necessary to work around delayed element removal when using the animations
```
We should also clarify whether this happens in Safari and IE or in all browsers.

theodorejb replied:
Clarified.

AndrewKushnir reviewed (changes requested):
Hi @theodorejb, thanks for additional changes and the great news that the `ngZone.onStable` works as expected. I think we are very close!

I did a quick review and left a few comments. The main one is around the reconciliation process, where the `writeValue` is always called. I'm wondering if it can be conditional (i.e. do that only when there is a discrepancy). Also the `selectedIndex` logic was removed from the `writeValue` function, which might be breaking/unexpected in some cases. In ideal case we should try to keep the `writeValue` as is + have a reconciliation logic that is triggered *only* when there is a discrepancy. In this situation in over 80% cases we'll keep the current logic and only in 20% (Safari and IE users), additional reconciliation will be triggered.

Thanks again for your time and efforts in fixing this problem! 👍

jelbourn commented on packages/forms/test/value_accessor_integration_spec.ts:
```diff
@@ -304,6 +304,35 @@ import {dispatchEvent} from '@angular/platform-browser/testing/src/browser_util'
              expect(buffalo.nativeElement.selected).toBe(true);
            }));
 
+        it('when new options are added but not selected', fakeAsync(() => {
```
optional since other tests were already like this: it would be great to name this test to more clearly describe the scenario. I'm thinking something like
```ts
it('should not select options when they are added after the select renders')
```
Maybe also adding a comment in the test that points to the relevant GitHub issue(s)

theodorejb replied:
Changed and added comment referencing GitHub issue.

theodorejb commented on packages/forms/src/directives/select_control_value_accessor.ts:
```diff
@@ -89,12 +89,14 @@ function _extractId(valueString: string): string {
   /** @internal */
   _idCounter: number = 0;
+  /** @internal */
+  _actualValue: string = '';
```
Done

AndrewKushnir commented on packages/forms/src/directives/select_control_value_accessor.ts:
```diff
@@ -123,7 +131,31 @@ export class SelectControlValueAccessor implements ControlValueAccessor {
+     * text nodes to the DOM, so it is necessary to set the value again after checking the view.
+     *
+     * This is also necessary to work around delayed element removal when using the animations
+     * module. Otherwise when a selected option is removed (so no option matches the ngModel
```
Minor suggestion to mention that the second problem actually happens in all browsers (since the first paragraph mentions Safari and IE specifically).
```suggestion
     * module (in all browsers). Otherwise when a selected option is removed (so no option matches the ngModel
```

theodorejb replied:
Previously it was mentioned at the end of the paragraph, but I moved it to the suggested location to make it more visible.

AndrewKushnir commented on packages/forms/src/directives/select_control_value_accessor.ts:
```diff
@@ -139,6 +171,7 @@ export class SelectControlValueAccessor implements ControlValueAccessor {
     }
     const valueString = _buildValueString(id, value);
     this._renderer.setProperty(this._elementRef.nativeElement, 'value', valueString);
+    this._actualValue = this._elementRef.nativeElement.value;
```
I'd like to propose a small refactoring to get rid of the additional `_actualValue` field and use the value from the `value` field directly.

We can extract the logic that updates the DOM into a separate function, like this one:
```
  /**
   * @internal
   */
  _updateDOMValue(selectedIndex: string|null, value: string) {
      if (id == null) {
        this._renderer.setProperty(this._elementRef.nativeElement, 'selectedIndex', -1);
      }
      this._renderer.setProperty(this._elementRef.nativeElement, 'value', value);
  }
```

Then we can use this function in both `writeValue` and `ngAfterViewChecked` methods:
```
  writeValue(value: any): void {
    this.value = value;
    const id: string|null = this._getOptionId(value);
    const valueString: string = _buildValueString(id, value);
    this._updateDOMValue(id, valueString);
  }
```

and
```
  ngAfterViewChecked() {
    // ...
    this._ngZone.onStable.pipe(take(1)).subscribe(() => {
      const id: string|null = this._getOptionId(this.value);
      const valueString: string = _buildValueString(id, this.value);
      if (this._elementRef.nativeElement.value !== valueString) {
        this._updateDOMValue(id, valueString);
      }
    });
  }
```

theodorejb replied:
I looked into this before, but it turns out that it doesn't work. The result of `_buildValueString` doesn't always match the actual value that's stored in the select element's `value` property. This breaks the placeholder test, for example.

And even if this approach did work, at this point we'd basically be duplicating most of the `writeValue` method, and losing whatever performance benefit we had by skipping the call to `_getOptionId`. We might as well put the check in `writeValue` instead to keep things simpler. But is there a benefit to checking the value before setting it anyway? It should be just as fast or faster to allow the browser to optimize setting the value when it hasn't changed.

I changed the implementation back to just call `writeValue` without additional checks, since I believe this is the simplest and most efficient approach that can work.

AndrewKushnir commented on packages/forms/test/value_accessor_integration_spec.ts:
```diff
@@ -304,6 +304,36 @@ import {dispatchEvent} from '@angular/platform-browser/testing/src/browser_util'
              expect(buffalo.nativeElement.selected).toBe(true);
            }));
 
+        it('should not select options added after the select renders', fakeAsync(() => {
```
It looks like the `BrowserAnimationsModule`-specific test was removed? I think we should bring it back (as it verifies the fix). If there is a problem with test setup, I can help have a look at it, please let me know.

theodorejb replied:
Okay, I added this test back now in a second commit. If it needs to be with the forms tests instead of animation tests, I will need someone to help with this, since I wasn't able to figure out how to make the forms test components render with `DOMTestComponentRenderer`.

AndrewKushnir reviewed (commented):
Hi @theodorejb, thanks again for working on the fix.

I've added a couple comments, please have a look when you get a chance. I think the next steps after addressing the remaining feedback would be:
- adding the `BrowserAnimationsModule`-related test
- run tests in Google's codebase (I will handle that)

Based on the presubmit results in Google's codebase we'll see if any adjustments to this PR are required and what the merge strategy would be (master-only or master+patch).

Thank you.

mary-poppins commented:
You can preview 649d026 at https://pr23784-649d026.ngbuilds.io/.
You can preview e020869 at https://pr23784-e020869.ngbuilds.io/.
You can preview 27e3dbd at https://pr23784-27e3dbd.ngbuilds.io/.
You can preview acf3ba7 at https://pr23784-acf3ba7.ngbuilds.io/.
You can preview 8da4acc at https://pr23784-8da4acc.ngbuilds.io/.
You can preview b25a60d at https://pr23784-b25a60d.ngbuilds.io/.
You can preview 048f566 at https://pr23784-048f566.ngbuilds.io/.
You can preview 842e46a at https://pr23784-842e46a.ngbuilds.io/.
You can preview a763b42 at https://pr23784-a763b42.ngbuilds.io/.
You can preview ebcf267 at https://pr23784-ebcf267.ngbuilds.io/.
You can preview 98e185c at https://pr23784-98e185c.ngbuilds.io/.
You can preview ee02bc3 at https://pr23784-ee02bc3.ngbuilds.io/.
You can preview fd6b449 at https://pr23784-fd6b449.ngbuilds.io/.
You can preview 503b6ff at https://pr23784-503b6ff.ngbuilds.io/.
You can preview 46b7355 at https://pr23784-46b7355.ngbuilds.io/.
You can preview 96c1ab3 at https://pr23784-96c1ab3.ngbuilds.io/.
You can preview 5d6eb8a at https://pr23784-5d6eb8a.ngbuilds.io/.
You can preview 5720e87 at https://pr23784-5720e87.ngbuilds.io/.
You can preview db3351b at https://pr23784-db3351b.ngbuilds.io/.
You can preview 39403ee at https://pr23784-39403ee.ngbuilds.io/.
You can preview dde1d2b at https://pr23784-dde1d2b.ngbuilds.io/.
You can preview d83e8a2 at https://pr23784-d83e8a2.ngbuilds.io/.
You can preview e412b10 at https://pr23784-e412b10.ngbuilds.io/.
You can preview d3793a2 at https://pr23784-d3793a2.ngbuilds.io/.
You can preview 745d3c4 at https://pr23784-745d3c4.ngbuilds.io/.
You can preview 32b946f at https://pr23784-32b946f.ngbuilds.io/.
You can preview 60bacb5 at https://pr23784-60bacb5.ngbuilds.io/.
You can preview f4db8e3 at https://pr23784-f4db8e3.ngbuilds.io/.
You can preview 94f36ba at https://pr23784-94f36ba.ngbuilds.io/.
You can preview 315815f at https://pr23784-315815f.ngbuilds.io/.
You can preview 2836b7a at https://pr23784-2836b7a.ngbuilds.io/.
You can preview 591228c at https://pr23784-591228c.ngbuilds.io/.
You can preview 9d857e0 at https://pr23784-9d857e0.ngbuilds.io/.
You can preview 7972caf at https://pr23784-7972caf.ngbuilds.io/.
You can preview 25bedd2 at https://pr23784-25bedd2.ngbuilds.io/.
You can preview b763e60 at https://pr23784-b763e60.ngbuilds.io/.
You can preview 2220421 at https://pr23784-2220421.ngbuilds.io/.

AndrewKushnir commented:
Hi @theodorejb,

Thanks for adding `BrowserAnimationsModule`-related test, I'll try to move it to the `forms` package and let you know.

I've generated the angular.io preview, which is available [here](https://pr23784-2220421.ngbuilds.io/api/forms/SelectControlValueAccessor) and it looks like the `ngAfterViewChecked` method requires some docs, [similar to this one](https://github.com/angular/angular/blob/master/packages/forms/src/directives/ng_form.ts#L140-L146). Could you please add them?

I also started [tests in Google's codebase](http://test/OCL:316938191:BASE:316938256:1592421196587:434eedae) and will let you know the result once I have it.

Thank you.

theodorejb commented:
@AndrewKushnir I added a description to the `ngAfterViewChecked` method now.

AndrewKushnir commented:
Hi @theodorejb, I've moved the `BrowserAnimationsModule`-specific test to the `forms` package, [the diff can be found here](https://gist.github.com/AndrewKushnir/a67d5754228af6fdf1015cea78075060). I also verified that the test fails without a fix from this PR. Please let me know if that works for you (btw, I think you can merge commits into one). Thank you.

theodorejb commented:
Hi @AndrewKushnir,

After applying that diff, the test still fails for me locally with the same type of error I was getting before:

> Error: The selector "#root38" did not match any elements

I pushed this change as a separate commit, so we can see if it also fails in CI and maybe you will be able to debug the issue more easily.

Once this issue is resolved I can squash the commits together before the PR is merged.

AndrewKushnir commented:
@theodorejb I think we should run it only in a browser (where the problem can actually be reproduced), i.e. add the following condition at the very beginning of the test (similar to what you have in other tests added in this PR):
```
if (isNode) return;
```
I've ran that test using the following command:
```
yarn bazel test packages/forms/test:test_web
```
So I think once you add the `isNode` check, the CI should be "green".
Please let me know if that resolved the issue.

theodorejb commented:
@AndrewKushnir Indeed, the `isNode` check resolved the issue.

I've squashed the commits together now, so this should be good to go if CI is green.

Thanks so much!

AndrewKushnir commented:
Hi @theodorejb, thanks for updating the PR.

Tests in Google's codebase indicated a couple places that didn't work well with this change. I will perform further investigation and let you know. Note: investigation may take a few days to fully complete.

AndrewKushnir commented:
Hi @theodorejb, just want to give a quick update: we've adapted the code in a couple places in Google's codebase, but there is a couple more that require further investigation (which is in progress). I will keep this thread updated. Thank you.

theodorejb commented:
@AndrewKushnir Any update on the progress? I just want to make sure this isn't forgotten. Let me know if there's anything I need to do.

devversion commented on packages/forms/src/directives/select_control_value_accessor.ts:
```diff
@@ -123,7 +123,33 @@ export class SelectControlValueAccessor implements ControlValueAccessor {
+     * ngModel.
+     */
+
+    this._ngZone.onStable.pipe(take(1)).subscribe(() => {
```
I think this could cause another round of change detection if someone uses the ZoneJS rx patches. We could get around it by just subscribing outside of the Angular zone.

theodorejb replied:
Okay, I changed this to subscribe outside the Angular zone, and verified that all the tests still pass.

devversion replied:
thanks!

devversion reviewed (dismissed):
LGTM for dev-infra. One comment about the subscription being made in the Angular zone.

AndrewKushnir commented:
Hi @theodorejb,

> Any update on the progress? I just want to make sure this isn't forgotten. Let me know if there's anything I need to do.

As I mentioned in my previous messages, there is a couple places in Google's codebase that require additional investigation. Currently I'm waiting for a reply from one of the internal Google teams potentially affected by this change - they have a tricky use-case and repro. I will keep you updated once I have more information. Thank you.

theodorejb commented:
> Currently I'm waiting for a reply from one of the internal Google teams potentially affected by this change - they have a tricky use-case and repro.

@AndrewKushnir Were you able to get a reply from the team? This is still a really important issue for us to get fixed.

AndrewKushnir commented:
Hi @theodorejb, sorry for the delay in answering and thanks for rebasing this PR to keep it up-to-date.

We performed further investigation and had a couple discussions with the team internally. The original problem (that this PR tries to resolve) appears due to the mismatch between Angular internal state and the DOM state (i.e. an option becomes selected in the browser when in Angular no options were set as selected), because of the specific browser behavior that you mentioned in PR description. The change in this PR is a possible solution to this, but there are some potential risks of application code relying on current DOM value (or changing it via `Renderer2` for example), so resetting it in async way might introduce unexpected behavior for users.

We'll perform additional investigation to see if there is a way to keep the state in sync in that situation and I'll share more results, but I don't have any ETAs at this moment.

theodorejb commented:
@AndrewKushnir I'm struggling to understand what exactly would break that isn't already broken. If an application relies on a synchronous read of the current DOM value, that won't be affected by this PR. If an application is making a synchronous write to the DOM value, that's already broken in the cases described in the PR description.

When `BrowserAnimationsModule` is imported, there's already an async DOM update causing a mismatch between the DOM and `ngModel` state. In theory maybe someone could have async code that relies on the ngModel state not matching the DOM, but this seems incredibly unlikely. I don't think it makes sense to hold back a fix for known bugs, that are definitely causing unexpected behavior for users, just because it is theoretically possible that someone could be relying on the bug (which in all likelihood is not the case).

If it turns out that this fix does actually cause issues for anyone, at that point maybe it would make sense to revert it and seek a different solution. Otherwise it seems like an unnecessary waste of time and resources for everyone. Of course, if the Angular team finds a cleaner way to solve these issues in the future, merging this PR now will not prevent that. If anything the tests added in this PR will make it easier to improve the implementation in the future without breaking things.

xonuzofa commented:
Any progress on this? Been a few months!

theodorejb commented:
> Any progress on this? Been a few months!

I really wish there was progress. What I did for my app is set up patch-package to apply the changes to select_control_value_accessor.ts in this PR to node_modules/@angular/forms/fesm2015/forms.js. I had to do this since select dropdowns are completely broken for my customers using Safari without this fix.

I hope this can be addressed in an official Angular release soon since it's a pain to update my patch whenever the file is modified in new versions of Angular.

xonuzofa commented:
Same.  Every time the data dynamically is updated, it changes my selects to the default value="" option.  If I use the animation module, it is really messed up.

arturovt commented on packages/forms/src/directives/select_control_value_accessor.ts:
```diff
@@ -126,7 +126,36 @@ export class SelectControlValueAccessor implements ControlValueAccessor {
+     */
+
+    // subscribe outside Angular zone to avoid triggering extra round of change detection
+    this._ngZone.runOutsideAngular(() => {
```
nit: `onStable` already emits events outside of the zone:
https://github.com/angular/angular/blob/dc9fd1aaefa9cf2a019e293581117a8e84ae277c/packages/core/src/zone/ng_zone.ts#L328

theodorejb replied:
Thanks. I rebased the PR and removed the extra `runOutsideAngular()` call.

dylhunn commented:
@theodorejb Now that we have a bit more manpower on Forms (that would be me!), I can run the g3 presubmit again and see if there's still breakage. Has this been rebased recently?

dylhunn commented:
Hi @theodorejb, thanks for working on the changes in this PR! I had a quick chat about this with @AndrewKushnir, and the problem with this fix is that changing DOM value happens asynchronously and that can lead to subtle issues as described in [this comment](https://github.com/angular/angular/pull/23784#issuecomment-692342393). We think the best approach would be to update the DOM value once the rendering is complete. Currently there is no lifecycle hook that for that, and we'd need a project proposal to add one. If there is a decision to add such lifecycle hook, the fix in this PR can leverage that and we should be able to land the change.

This is certainly a disappointing outcome, since this is a major performance issue and you put in a lot of effort on this. Sorry that it turned out to be blocked on a major addition like this :/

theodorejb commented:
@dylhunn Is someone going to make a project proposal to add the needed lifecycle hook? I'm not sure how to go about that myself. I solved the issue in our apps by applying the patch in this PR using `patch-package`.

I'm surprised fixing this hasn't had a higher priority, given that "Speed and Performance" is hailed as one of Angular's core features, but select bindings currently perform far worse than in competing frameworks.

AndrewKushnir commented:
Hi @theodorejb,

> Is someone going to make a project proposal to add the needed lifecycle hook? I'm not sure how to go about that myself. I solved the issue in our apps by applying the patch in this PR using patch-package.

Yes, Angular team will publish a proposal once we perform some initial investigation to assess the implications and affected subsystems.

dylhunn commented:
@theodorejb Would you be able to rebase this again? I will run a presubmit so we have real data on whether any project is broken in practice.

theodorejb commented:
@dylhunn It's rebased again now.

dylhunn commented:
Thanks @theodorejb, I am running a fresh presubmit now. [link](http://test/OCL:441889509:BASE:443159036:1650481906731:9c1c974f)

theodorejb commented:
@dylhunn Did the presubmit uncover any issues that would be difficult to fix?

dylhunn commented:
> Did the presubmit uncover any issues that would be difficult to fix?

Nothing obvious! There might be individual breakages in a global run, but nothing huge or it would show up in the standard presubmit.

alfaproject commented:
OOC would an opt-in flag be an option for change in behaviour?

theodorejb commented:
Since v15 is in development now, has a decision been made about how to fix the bugs targeted by this PR? I noticed an issue was filed for providing a new lifecycle hook - is there a plan to implement that in v15?

dylhunn commented:
@theodorejb Unfortunately, there has been no progress on this. After some discussion, we feel that the right approach is to design and implement the new lifecycle hook, tracked in #47432. We have not budgeted time for that project yet.

theodorejb commented:
What's the status of the new lifecycle hook? Has any progress been made implementing it in v16? If there isn't a concrete plan to add it in the next version, can this pull request at least be merged so select dropdowns with `ngValue` don't remain broken in the meantime? That wouldn't prevent still adding the lifecycle hook later.

atscott commented on packages/forms/src/directives/select_control_value_accessor.ts:
```diff
@@ -146,9 +139,7 @@ export class SelectControlValueAccessor extends BuiltInControlValueAccessor impl
-
-    // subscribe outside Angular zone to avoid triggering extra round of change detection
-    this._ngZone.onStable.pipe(take(1)).subscribe(() => {
+    afterRender(() => {
```
I believe you want `afterNextRender` here. `afterRender` will run after _every_ change detection but you only want this to run a single time.

theodorejb replied:
With `afterNextRender`, wouldn't it have to be called every time an option is added or removed, rather than once in the component constructor?

atscott replied:
> With afterNextRender, wouldn't it have to be called every time an option is added or removed, rather than once in the component constructor?

Yes, I missed the change of the code location. Maybe you still want it in `afterViewChecked`. `afterRender` will certainly run more times than what you had in previous versions.

theodorejb commented:
It looks like tests are failing with the following error:

> NG0203: afterNextRender() can only be used within an injection context such as a constructor, a factory function, a field initializer, or a function used with `runInInjectionContext`.

I'm not sure how to fix this.

JeanMeche commented:
`afterNextRender` is expected to run in an [injection context](https://angular.io/guide/dependency-injection-context). 

You should wrap the call of `afterNextRender` in an `runInInjectionContext(injector, () => {  afterNextRender(() => {...})  })`

theodorejb commented:
The same error is still occurring. I'm not sure if `EnvironmentInjector` is the correct injector to use or if something else is incorrect.

JeanMeche commented:
It's a different line that is throwing. In the unit test `directives_spec.ts:72:34` the `SelectControlValueAccessor` is created outsite of DI. 

Looks the test will need to rely on DI too !

AndrewKushnir commented on packages/forms/src/directives/select_control_value_accessor.ts:
```diff
@@ -117,6 +117,35 @@ export class SelectControlValueAccessor extends BuiltInControlValueAccessor impl
+     * from the DOM, the browser would change the select value to that of the first option, even
+     * though it doesn't match the ngModel (issue #18430).
+     */
+    afterNextRender(() => {
+      this.writeValue(this.value);
+    }, {injector: this._injector});
```
We could wrap this code into the `runInInjectionContext` function to resolve the issue identified by CI:
```suggestion
    runInInjectionContext(this.injector, () => {
      afterNextRender(() => {
        this.writeValue(this.value);
      });
    });
```

theodorejb replied:
Now there's an error that the call to `inject` on line 120 (to initialize `this._injector`) must be called from an injection context. How is `this.injector` in your example supposed to be initialized?

AndrewKushnir replied:
In my example the `this.injector` is initialized in the same way you have it in your PR.

The problem with the `inject()` in this class is a test setup issue: those failing tests are not compatible with the `inject()` function, since directive instances are created via `new` ([see here](https://github.com/angular/angular/blob/main/packages/forms/test/directives_spec.ts#L85-L89 )). To fix this, those failing tests can be wrapped into a similar function (the `TestBed.runInInjectionContext` one) to establish an injection context. Here is an example:

```
diff --git a/packages/forms/test/directives_spec.ts b/packages/forms/test/directives_spec.ts
index 1bda48e1b5..114abfd01a 100644
--- a/packages/forms/test/directives_spec.ts
+++ b/packages/forms/test/directives_spec.ts
@@ -7,7 +7,7 @@
  */
 
 import {SimpleChange} from '@angular/core';
-import {fakeAsync, flushMicrotasks, tick} from '@angular/core/testing';
+import {fakeAsync, flushMicrotasks, TestBed, tick} from '@angular/core/testing';
 import {AbstractControl, CheckboxControlValueAccessor, ControlValueAccessor, DefaultValueAccessor, FormArray, FormArrayName, FormControl, FormControlDirective, FormControlName, FormGroup, FormGroupDirective, FormGroupName, NgControl, NgForm, NgModel, NgModelGroup, SelectControlValueAccessor, SelectMultipleControlValueAccessor, ValidationErrors, Validator, Validators} from '@angular/forms';
 import {selectValueAccessor} from '@angular/forms/src/directives/shared';
 import {composeValidators} from '@angular/forms/src/validators';
@@ -69,10 +69,12 @@ class CustomValidatorDirective implements Validator {
         });
 
         it('should return select accessor when provided', () => {
-          const selectAccessor = new SelectControlValueAccessor(null!, null!);
-          expect(selectValueAccessor(dir, [
-            defaultAccessor, selectAccessor
-          ])).toEqual(selectAccessor);
+          TestBed.runInInjectionContext(() => {
+            const selectAccessor = new SelectControlValueAccessor(null!, null!);
+            expect(selectValueAccessor(dir, [
+              defaultAccessor, selectAccessor
+            ])).toEqual(selectAccessor);
+          });
         });
 
         it('should return select multiple accessor when provided', () => {
@@ -83,9 +85,12 @@ class CustomValidatorDirective implements Validator {
         });
 
         it('should throw when more than one build-in accessor is provided', () => {
-          const checkboxAccessor = new CheckboxControlValueAccessor(null!, null!);
-          const selectAccessor = new SelectControlValueAccessor(null!, null!);
-          expect(() => selectValueAccessor(dir, [checkboxAccessor, selectAccessor])).toThrowError();
+          TestBed.runInInjectionContext(() => {
+            const checkboxAccessor = new CheckboxControlValueAccessor(null!, null!);
+            const selectAccessor = new SelectControlValueAccessor(null!, null!);
+            expect(() => selectValueAccessor(dir, [checkboxAccessor, selectAccessor]))
+                .toThrowError();
+          });
         });
 
         it('should return custom accessor when provided', () => {
```

theodorejb commented on packages/forms/src/directives/select_control_value_accessor.ts:
```diff
@@ -117,6 +117,37 @@ export class SelectControlValueAccessor extends BuiltInControlValueAccessor impl
+     * from the DOM, the browser would change the select value to that of the first option, even
+     * though it doesn't match the ngModel (issue #18430).
+     */
+    runInInjectionContext(this._injector, () => {
```
What would queue the hook only when the value is dirty? Would I have to update a flag on the class whenever an option value is set or destroyed, and then only queue the hook in `ngAfterViewChecked` if the flag is set?

atscott replied:
Could the afterNextRender be called in the input setter or ngOnChanges instead?

theodorejb replied:
@atscott I moved queuing the hook to the input setters, which seems to work the same as when I did it in `ngAfterViewChecked`.

theodorejb commented:
There's still one test failing, and I'm not sure why:

> value accessors select controls in template-driven forms when options are removed
> Message:  Expected '1: Object' not to equal '1: Object'.

dylhunn commented:
Some initial thoughts: If the value is updated via `set value` twice in the same CD cycle, wouldn't the second value get dropped (because an update is already queued)?

JoostK commented:
> Some initial thoughts: If the value is updated via `set value` twice in the same CD cycle, wouldn't the second value get dropped (because an update is already queued)?

That should result in dropping the first update, not the second; flushing the write to the DOM in the `afterNextRender` hook always reads the current value, not the value that was set when the hook was requested.

dylhunn reviewed (approved):
reviewed-for: fw-forms

I will run a TAP global presubmit tonight. If it's green, we should target 17.0.

dylhunn commented:
Also @theodorejb -- your final commit message is failing lint because it needs a body >20 chars apparently.

theodorejb commented:
@dylhunn Thanks! I rebased now and squashed the last 4 commits with an updated message.

dylhunn commented:
@theodorejb Sorry about this, mind rebasing one more time? My TGP failed to get scheduled (we've been having capacity restrictions...) and I need to try it again.

theodorejb commented:
@dylhunn It's rebased now.

dylhunn commented:
TGP: [google link](https://test.corp.google.com/OCL:568715636:BASE:568724937:1695785048477:98d48b15)

theodorejb commented:
@dylhunn Do the broken tests indicate a problem with this patch? If so, could a similar failing test be added to Angular itself? I'm not sure how I could debug those errors without being able to reproduce them myself.

thePunderWoman commented:
This would be great to land, but unfortunately the amount of effort involved for us to safely get it submitted is too high. So we're going to close this, but thank you for your contribution. We appreciate you!

theodorejb commented:
Despite the countless hours I put into fixing this, I don't personally care anymore, since I moved my projects to Svelte which has been [far easier to contribute to](https://github.com/sveltejs/svelte/issues?q=state%3Aclosed%20is%3Apr%20author%3Atheodorejb). But I feel bad for all the Angular users who are stuck with these apparently unresolvable bugs.

atscott commented:
The history here is also quite long, with extensive involvement from the team over the years. I cannot see each iteration of the change, but concerns with the changes cannot be invalidated. We also have to balance priorities and cannot dedicate infinite amounts of time to any one issue.

I gave this another look after it was closed yesterday and verified real issues with the current iteration of the changes in the internal test failures. I hesitate to even mention this based due to the history here: I think I have addressed those issues and am truly hopeful we can finally resolve this, but I cannot make any promises.
